### ISSUE TITLE
Removing the "has_nulls" option from to_parquet, or set default to True?

### ISSUE THREAD

**Issue by wesm** [dask/dask#2177]
In dask.dataframe.io.to_parquet: 

```
    has_nulls : bool, list or None
        Specifies whether to write NULLs information for columns. If bools,
        apply to all columns, if list, use for only the named columns, if None,
        use only for columns which don't have a sentinel NULL marker (currently
        object columns only).
```

I am looking in the logic in:

https://github.com/dask/fastparquet/blob/master/fastparquet/writer.py#L626

My understand of this is that the default behavior is to _not_ write nulls in the Parquet definition levels (i.e. `RepetitionType:REQUIRED`) for float32/64 types. I'm concerned this is a dangerous default setting because the Parquet files written will not be readable in general by other systems that use Parquet files. Would you be open to removing this option altogether (respecting pandas's NULL semantics) or changing its default value to `True`?

cc @martindurant

**Comment by wesm** [comment]
Bumping this issue. If files are being written with nonzero probability of being read by Spark, Hive, or Impala, then the default options are producing files whose results in those systems may be incorrect. 

If using `has_nulls=False` yields better performance in a dask-to-dask roundtrip, that seems OK, but that should be something that users are explicitly opting in to. I think you also should clearly document that this will produce incompatible Parquet files.

**Comment by martindurant** [comment]
I don’t see how it can be incorrect to save the value NaN as opposed to NULL in a parquet file, when the input is a pandas float column in which, as far as I know, there is no difference between the two.

**Comment by wesm** [comment]
The NaN values in pandas are semantically NULL; when you write them as NaN they are no longer considered to be NULL by other systems. So

`pd.isnull(value)`

will yield different results from `ISNULL(value)` in Spark/Hive/Impala/Presto/Drill/etc.

**Comment by martindurant** [comment]
"The NaN values in pandas are semantically NULL" - so how does one achieve a semantic NaN?

**Comment by wesm** [comment]
You cannot -- NaN is null:

```
In [1]: import pandas as pd

In [2]: s = pd.Series([-1, -1, -1])

In [3]: import numpy as np

In [4]: nans = np.sqrt(s)
/home/wesm/anaconda3/envs/arrow-test/bin/ipython:1: RuntimeWarning: invalid value encountered in sqrt
  #!/home/wesm/anaconda3/envs/arrow-test/bin/python

In [5]: nans
Out[5]: 
0   NaN
1   NaN
2   NaN
dtype: float64

In [6]: nans.isnull()
Out[6]: 
0    True
1    True
2    True
dtype: bool
```

**Comment by mrocklin** [comment]
I think that the following is a simple and motivational use case in favor of treating nans/nulls dilligently (what @wesm describes).  I haven't thought about this topic enough to have a strong opinion, but I think that this example might be useful.

Imagine you read the following CSV file

```
a,b
1,1
2,2
3,NA
4,4
```

Pandas treats the missing values here properly as missing values

```python
>>> df = pd.read_csv('myfile.csv')
>>> df.b.sum()
7
```

But now we move the data to parquet and then read with Spark and try the same thing:

```python
>>> df.agg({'b': 'sum'})
NaN
```

So either Spark would need to interpret NaN's as Nulls (which I suspect is unlikely and probably incorrect from their perspective) or we are transforming the data so that it produces different results on different tools.

**Comment by wesm** [comment]
Overmore, I believe that RepetitionType::REQUIRED should be avoided altogether, since if you append data to a dataset written initially by Dask, you may have a schema conflict (many systems use OPTIONAL exclusively).

**Comment by cpcloud** [comment]
@martindurant

It's incorrect to interpret `NaN` as IEEE `NaN` when using pandas. They *must* be interpreted as `NULL`. The choice was made a long time ago, for better or worse and there's no going back. Future systems will correct this inconsistency. Until then, if you choose to keep the standard `NaN` meaning when writing to parquet files you will create files that are not compatible with the systems that use parquet the most (Impala and Spark).

Here's a concrete example using Impala:

* Create a simple Dask DataFrame from a pandas DataFrame:
```
In [1]: import dask.dataframe as dd, numpy as np, pandas as pd

In [2]: df = pd.DataFrame({'a':[1,2,3],'b':[1.0,np.nan,3.0]})

In [3]: ddf = dd.from_pandas(df, npartitions=1)

In [4]: ddf.to_parquet?

In [5]: ddf.to_parquet('proper_foo.parquet', write_index=False, has_nulls=True)
```

* Fire up impala and create an external table from the above:

```
[localhost:21000] > create external table proper_foo_t (a int, b double)
stored as parquet location '/user/hive/warehouse/proper_foo';
```

* Try to select some data from this table:

```
[localhost:21000] > select a from proper_foo_t;
Query: select a from proper_foo_t
Query submitted at: 2017-04-12 14:04:03 (Coordinator: http://impala:25000)
Query progress can be monitored at: http://impala:25000/query_plan?query_id=e49da4ce39b675e:9d9b3bc200000000
WARNINGS:
File 'hdfs://172.17.0.2:9000/user/hive/warehouse/proper_foo/part.0.parquet' has an incompatible Parquet schema for column 'default.proper_foo_t.a'. Column type: INT, Parquet schema:
required int64 a [i:1 d:0 r:0]



File 'hdfs://172.17.0.2:9000/user/hive/warehouse/proper_foo/part.0.parquet' has an incompatible Parquet schema for column 'default.proper_foo_t.a'. Column type: INT, Parquet schema:
required int64 a [i:1 d:0 r:0]
```

The choice to use `REQUIRED` *at all* is fundamentally incorrect if you expect this data to be used by any other system besides Dask.

**Comment by wesm** [comment]
Here's an example of writing a DataFrame with nulls with the default options and then querying it with Impala:

Write the file:

```python
import dask.dataframe as dd, numpy as np, pandas as pd
df = pd.DataFrame({'key': [u'a', u'b', u'c'] * 100,
                   'value': np.random.randn(300)})
df['value'][::4] = np.nan
ddf = dd.from_pandas(df, npartitions=1)
​
ddf.to_parquet('proper_foo.parquet', write_index=False)
```

Do a groupby with pandas

```python
df.groupby('key').value.sum()
Out[11]:
key
a    13.728379
b    -0.028474
c    -2.610579
Name: value, dtype: float64
```

Create an EXTERNAL TABLE in Impala referencing the new Parquet file:

```python
import ibis
import pandas as pd
hdfs = ibis.hdfs_connect('localhost', port=5070)
con = ibis.impala.connect('localhost', port=21050, hdfs_client=hdfs)
ibis.options.verbose = True

path = '/tmp/fastparquet-test'
hdfs.mkdir(path)

hdfs.put('/tmp/fastparquet-test/0.parq', 'proper_foo.parquet/part.0.parquet')
```

Create a table referencing the file:

```python
>>> pf = con.parquet_file(path)
SHOW DATABASES LIKE '__ibis_tmp'
CREATE EXTERNAL TABLE __ibis_tmp.`__ibis_tmp_524b11f116a04d29add65f1bd9bfee8c`
LIKE PARQUET '/tmp/fastparquet-test/0.parq'
STORED AS PARQUET
LOCATION '/tmp/fastparquet-test'
```

Now do the same groupby, which yields NaN results:

```python
cur = con.con.execute('select key, sum(value) from __ibis_tmp.`__ibis_tmp_524b11f116a04d29add65f1bd9bfee8c` group by key')
cur.fetchall()
select key, sum(value) from __ibis_tmp.`__ibis_tmp_524b11f116a04d29add65f1bd9bfee8c` group by key
Out[18]:
[('b', nan), ('a', nan), ('c', nan)]
```

For the sake of compatibility, in Dask I strongly recommend:

* Always use OPTIONAL repetition type
* Write with `has_nulls=True`

I think it's OK if you want to have a special option for fastparquet users to disable compatibility but give better performance, but it should not be the default.

**Comment by wesm** [comment]
@cpcloud your example has a mistake because the type of `a` is bigint (int64) not int

**Comment by cpcloud** [comment]
Removed, the example works with proper types.

**Comment by mrocklin** [comment]
Checking in on the status here.  @martindurant did you have any thoughts?

**Comment by martindurant** [comment]
Sorry for my relative silence on this issue. Here are a couple of further thoughts:

- from a more scientist point of view, there is an argument that nan should mean nan as the result of an invalid calculation and not the same as null or missing data. Of course, this is not what pandas normally does, and I'm sure the reasons for that were thorough (had to choose one or the other).
- I note that spark *does* have a concept of nan, which is not the same as null, and if we have has_nulls=True by default, the user will never make them
- the performance cost on reading a float column with nulls in is about 30% in my benchmarks. There is a tradeoff between compatibility and performance, which depends on the user's purpose. There are other similar [compatibility options](http://fastparquet.readthedocs.io/en/latest/details.html) in a similar category and it seems that the implementation of the parquet options is spotty in the various reader frameworks. (btw: can anyone send me parquet data containing delta-encoded ints??)

I am OK to say that performance comes second to compatibility in general, as long as the loss is not too big, if we are conscious that that is the decision being made.

On the point that all schema fields should be OPTIONAL: this sounds pretty odd to me, no designer of a database table would agree that the primary key should be optional, even if in practice it never is.

**Comment by martindurant** [comment]
Ah, I misunderstood your earlier statement about appends, I had honestly not thought of people using different tools to append to the same dataset. Does that happen?

**Comment by wesm** [comment]
In the case of Hive or Impala, the Hive metastore allows you to create partitions that can point to any directory in HDFS -- it's not unusual for different ETL processes to dump data into a given logical dataset. See e.g. https://blog.cloudera.com/blog/2015/11/how-to-ingest-and-query-fast-data-with-impala-without-kudu/

**Comment by pitrou** [comment]
@wesm' arguments sound reasonable to me. If it's important that people be able to choose performance presets easily, perhaps this should be handled by a dialect or compatibility setting in fastparquet, so that users don't have to bother about individual configuration options such as `has_nulls`?

**Comment by martindurant** [comment]
OK, I'm convinced, I'll make the change. This will be for has_nulls=True for both fastparquet and dask's to_parquet. What would be a reasonable name for the current behaviour (which is no longer default, and so should not be None): "fast_but_incompatible" ?

**Comment by mrocklin** [comment]
I recommend naming after the semantics like nan_as_null or has_nulls,
rather than performance adjectives like fast.

**Comment by wesm** [comment]
It seems like this is a candidate for a pass-through parameter to the engine. I'd to provide the option to write files with pyarrow also

**Comment by wesm** [comment]
Can we change this default option in Dask? I tried in https://github.com/dask/dask/commit/68f9e417924a985c1f2e2a587126833c70a2e9f4 but it broke the test suite, so more work is required

**Comment by martindurant** [comment]
Yes, agree that this can be True (but please do not remove).
I thought this was done, but I cannot find it now. 
More changes will come soon to the fastparquet side following the pandas schema spec, but feel free to change this immediately.

**Comment by mrocklin** [comment]
Has this been resolved?

**Comment by martindurant** [comment]
This is also in https://github.com/dask/dask/pull/2365 , which I now see has a py2 issue yet to be resolved; I'll get on that.

**Comment by jcrist** [comment]
Fixed in #2365 - the default setting of in `to_parquet` is now `has_nulls=True`. Closing.
